### ISSUE TITLE
Intermittent 401 Unauthorized errors

### ISSUE THREAD

**Issue by markbao** [burnash/gspread#373]
Sometimes works, sometimes doesn't. Restarting the app helps. Could this be because there is a long lag time between an authorization of a credential with `gspread.authorize` and then taking actions on the session?

```
HTTPError: 401: <HTML>
<HEAD>
<TITLE>Unauthorized</TITLE>
</HEAD>
<BODY BGCOLOR="#FFFFFF" TEXT="#000000">
<H1>Unauthorized</H1>
<H2>Error 401</H2>
</BODY>
</HTML>
Traceback (most recent call last):
  File "/home/mark/timeobserver/lib/python2.7/site-packages/flask/app.py", line 1836, in __call__
    return self.wsgi_app(environ, start_response)
  File "/home/mark/timeobserver/lib/python2.7/site-packages/flask/app.py", line 1820, in wsgi_app
    response = self.make_response(self.handle_exception(e))
  File "/home/mark/timeobserver/lib/python2.7/site-packages/flask/app.py", line 1403, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/home/mark/timeobserver/lib/python2.7/site-packages/flask/app.py", line 1817, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/mark/timeobserver/lib/python2.7/site-packages/flask/app.py", line 1477, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/mark/timeobserver/lib/python2.7/site-packages/flask/app.py", line 1381, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/home/mark/timeobserver/lib/python2.7/site-packages/flask/app.py", line 1475, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/mark/timeobserver/lib/python2.7/site-packages/flask/app.py", line 1461, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/mark/timeobserver/server.py", line 42, in toggleWork
    worksheet.append_row([session.timeStarted, datetime.datetime.now(), session.durationWorked, session.durationInterrupted, datetime.datetime.now() - session.timeStarted, session.interruptions])
  File "/home/mark/timeobserver/lib/python2.7/site-packages/gspread/models.py", line 525, in append_row
    self.add_rows(1)
  File "/home/mark/timeobserver/lib/python2.7/site-packages/gspread/models.py", line 507, in add_rows
    self.resize(rows=self.row_count + rows)
  File "/home/mark/timeobserver/lib/python2.7/site-packages/gspread/models.py", line 488, in resize
    feed = self.client.get_feed(self_uri)
  File "/home/mark/timeobserver/lib/python2.7/site-packages/gspread/client.py", line 256, in get_feed
    r = self.session.get(url)
  File "/home/mark/timeobserver/lib/python2.7/site-packages/gspread/httpsession.py", line 75, in get
    return self.request('GET', url, **kwargs)
  File "/home/mark/timeobserver/lib/python2.7/site-packages/gspread/httpsession.py", line 71, in request
    response.status_code, response.content))
```

gspread latest (0.3.0)

**Comment by dsinghvi** [comment]
I am getting the same issue now and was wondering if there is a solution?

**Comment by carloAI2** [comment]
Getting the same error, but TBH, my python script runs for hours, so I suspect that we have to reauthenticate every once in a while? Is there a way to make a persistent authorization connection, so we don't have to reauthenticate every N hours?

**Comment by rhemon** [comment]
I am also having the same issue, I am using this in my web app, which is working on google app engine. I can solve the issue for a while by re-deploying my web app, but it starts happening again after a while. I dont know if its about reauthenticating, since I tried that and there was no change. So please if you find solution, let me know.

**Comment by markbao** [comment]
I'm not sure if this will help in other cases, but the way that I solved this was to have authentication (with ` gspread.authorize`) happen right before I executed an action. So, instead of authenticating in the beginning of my script, and then using those details to execute an action, I authenticate before every action. I'm curious if there is a persistent way to do this as @carloAI2 says.

**Comment by rhemon** [comment]
@markbao , I am still facing the problem, even after using `gspread.authorize` at the beginning of every action. :'( why is it doing this to me?  if you know any other solution, please help me. (but thanks for the reply, I have been posting this problem in few forums: stackoverflow, google groups, but wasnt getting any response)

**Comment by burnash** [comment]
@rhemon What's your gspread version?

**Comment by rhemon** [comment]
@burnash 0.2.5

**Comment by burnash** [comment]
@rhemon Could you update to the latest version and try again?

**Comment by rhemon** [comment]
Looks like I get a new error if I update to the latest version : 
```
FeatureNotEnabledError: The Socket API will be enabled for this application once billing has been enabled in the admin console.
at create_connection (/base/data/home/apps/s~gowcanteen/1.394383211349002244/lib/requests/packages/urllib3/util/connection.py:67)
at _new_conn (/base/data/home/apps/s~gowcanteen/1.394383211349002244/lib/requests/packages/urllib3/connection.py:142)
at connect (/base/data/home/apps/s~gowcanteen/1.394383211349002244/lib/requests/packages/urllib3/connection.py:254)
at _validate_conn (/base/data/home/apps/s~gowcanteen/1.394383211349002244/lib/requests/packages/urllib3/connectionpool.py:814)
at _make_request (/base/data/home/apps/s~gowcanteen/1.394383211349002244/lib/requests/packages/urllib3/connectionpool.py:351)
at urlopen (/base/data/home/apps/s~gowcanteen/1.394383211349002244/lib/requests/packages/urllib3/connectionpool.py:578)
at send (/base/data/home/apps/s~gowcanteen/1.394383211349002244/lib/requests/adapters.py:403)
at send (/base/data/home/apps/s~gowcanteen/1.394383211349002244/lib/requests/sessions.py:585)
at request (/base/data/home/apps/s~gowcanteen/1.394383211349002244/lib/requests/sessions.py:475)
at get (/base/data/home/apps/s~gowcanteen/1.394383211349002244/lib/requests/sessions.py:487)
at request (/base/data/home/apps/s~gowcanteen/1.394383211349002244/lib/gspread/httpsession.py:68)
at get (/base/data/home/apps/s~gowcanteen/1.394383211349002244/lib/gspread/httpsession.py:76)
at get_spreadsheets_feed (/base/data/home/apps/s~gowcanteen/1.394383211349002244/lib/gspread/client.py:165)
at open (/base/data/home/apps/s~gowcanteen/1.394383211349002244/lib/gspread/client.py:81)
at __init__ (/base/data/home/apps/s~gowcanteen/1.394383211349002244/server.py:13)
at <module> (/base/data/home/apps/s~gowcanteen/1.394383211349002244/main.py:15)
```
Now I am only student at the moment and its possible for me to enable the billing on my own. Well this was something for my school, so will talk to my teacher and see if he is go for it or not. If there is any other way to solve this, please let me know. And many thanks for replying @burnash

**Comment by rhemon** [comment]
And also I am not sure if the latest version will solve that problem or not? I am very curious what is the problem and why is it occuring? Especially why does it run properly for a while after redeploying, and then after some hours (which isnt a constant amount) it starts giving HTTPError 401?

**Comment by bryzgaloff** [comment]
I have faced quite the same issue today: after a night running gspread refused to modify the spreasheet. Unfortunately I have no logs, because it was running not in main thread but I guess the problem is that Google session expires. Have anyone tried to recreate a credentials instance? I'm going to try it out.

My idea is to catch an HTTPError and recreate the credentials to attempt once again.

**Comment by rhemon** [comment]
if it works out share the code with me plz for  now all i do is re upload
the code every day

**Comment by bryzgaloff** [comment]
As [mentioned](https://cloud.google.com/storage/docs/json_api/v1/status-codes#401_Unauthorized) in Google Cloud Platform docs, one of the common reasons for 401 error is:
> The OAuth access token has expired and needs to be refreshed. This can be avoided by refreshing the access token early, but code can also catch this error, refresh the token and retry automatically.

That's why I decided to create such a class instead of gspread Worksheet class:
```
class WKS:
    def __init__(self):
        scope = ['https://spreadsheets.google.com/feeds']
        self.credentials = ServiceAccountCredentials.from_json_keyfile_name(secret.GOOGLE_JSON_FILENAME, scope)
        self._refresh_auth()

    def _refresh_auth(self):
        gc = gspread.authorize(self.credentials)
        self.wks = gc.open_by_key(secret.SPREADSHEET_KEY).sheet1

    def __getattr__(self, attr):
        if hasattr(self.wks, attr):
            try:
                return getattr(self.wks, attr)
            except gspread.exceptions.HTTPError as e:
                logging.warning('HTTPError: code = ' + str(e.code) + '\n' + e.args[0])
                self._refresh_auth()
                return getattr(self.wks, attr)
            except Exception as e:
                logging.error(str(e.args))
        else:
            return getattr(self, attr)
```
Frankly speaking, I haven't tried it in production but I hope it will help.

**Comment by rhemon** [comment]
thanks will try adding to my code

**Comment by bryzgaloff** [comment]
@rhemon, note please that I have edited the snippet in the comment a bit to prevent stack overflow.

**Comment by rhemon** [comment]
ok thanks for mentioning. i havent added it yet. plannig to do it tonight

**Comment by bryzgaloff** [comment]
Not that I have just changed code again. There was senseless handling :laughing: 

However, recreating the credentials doesn't help. I looked through the code and found the only instance, from where @markbao could get such an output. It's [here](https://github.com/burnash/gspread/blob/7627c98269e177f12715c71f38395db9268264a0/gspread/httpsession.py#L71). I'm still trying to find out the cause of an error.

**Comment by rhemon** [comment]
not sure exactly why but I have noticed that this error automatically goes
away after a while and then comes back and again goes away and the cycle
continues
and @anton I just realized that I am authorizing the credentials every time
before every action when requests come, so wouldnt the result be same if I
refresh the auth right after getting the error? so yeah i guess thats not
working. btw what version are you using? I am using an old version, since
latest version use a feature that requires billing to be enabled. So I am
curious whether the latest version fixes the issue or not, if its a fix for
sure I could talk to my teacher at school to enable billing for the project
and use the latest version of gspread

**Comment by bryzgaloff** [comment]
I'm currently using the latest version of gspread from pip: 0.4.1. The problem is certainly caused by credentials expiration after an hour script running. Right now I have found out an interesting thing, using my code from above: the `insert_row` method, which I'm using, worked fine after an hour of script running in the following chain of calls:
* `insert_row` (caught an HTTPError in outer try-block)
* `_refresh_auth()` (called from handler)
* `insert_row` (HTTPError again, caught in the inner try-block)
* `insert_row` (new attempt, succeed this time)

I have no idea how to explain it.

**Comment by bryzgaloff** [comment]
I have understood the reason: I have been still using the old (non-authorized properly) instance of worksheet, that's why only the new attempt succeed. So, I guess, proper reauthorizing is the solution for the issue.

**Comment by bryzgaloff** [comment]
I have tested the snippet and here is the final solution, that works for me (proxy class for worksheet):
```python
class WKS:
    def __init__(self):
        scope = ['https://spreadsheets.google.com/feeds']
        self.credentials = ServiceAccountCredentials.from_json_keyfile_name(secret.GOOGLE_JSON_FILENAME, scope)
        self._refresh_auth()

    def _refresh_auth(self):
        gc = gspread.authorize(self.credentials)
        self.wks = gc.open_by_key(secret.SPREADSHEET_KEY).sheet1

    def _decorate(self, method):
        def safe_method(*args, **kwargs):
            try:
                method(*args, **kwargs)
            except gspread.exceptions.HTTPError as e:
                # getattr is needed to get a new instance of self.wks
                getattr(self.wks, method.__name__)(*args, **kwargs)
        return safe_method

    def __getattr__(self, attr):  # doesn't shadow _refresh_auth and _decorate
        return self._decorate(getattr(self.wks, attr))
```

**Comment by rhemon** [comment]
Ok thanks.. added to my code, hope it will solve it , my code is some what
like this:
method:
    try:
         actions of spreadsheet
         count = 0
    except HTTPError:
          refresh_auth
          count += 1
          if (count =< 10):
                method
i added the count thing just incase (maybe its not really needed) so that
it doesnt keep iterating if the error doesnt go away

**Comment by bwatkin79** [comment]
Just wanted to give my experience with this error recently. I'm running the program through Python Anywhere.  A few days ago, my program was working fine, but I'm getting this error. Just a warning, I'm a beginner at programming, but hopefully my input will be of some use.

I have a workbook I've been using to test my program and a second workbook which will serve as the real destination for my data. I was able to send my data to the second workbook a few days ago, but now I'm getting the error. If I comment out the line for second workbook and remove the comment for the test workbook, my data pushes perfectly fine. Not sure why one workbook works but not the other.

This is the error I'm getting:

    Traceback (most recent call last):
      File "/home/bwatkin79/fangraphstest/fangraphsxpath.py", line 378, in <module>
        player['A'], player['B']])
      File "/home/bwatkin79/.local/lib/python3.5/site-packages/gspread/models.py", line 533, in append_row
        cell = self.cell(new_row, i)
      File "/home/bwatkin79/.local/lib/python3.5/site-packages/gspread/models.py", line 328, in cell
        self._cell_addr(row, col))
      File "/home/bwatkin79/.local/lib/python3.5/site-packages/gspread/client.py", line 269, in get_cells_cell_id_feed
        r = self.session.get(url)
      File "/home/bwatkin79/.local/lib/python3.5/site-packages/gspread/httpsession.py", line 75, in get
        return self.request('GET', url, **kwargs)
      File "/home/bwatkin79/.local/lib/python3.5/site-packages/gspread/httpsession.py", line 71, in request
        response.status_code, response.content))
    gspread.exceptions.HTTPError: 401: b'<HTML>\n<HEAD>\n<TITLE>Unauthorized</TITLE>\n</HEAD>\n<BODY BGCOLOR="#FFFFFF" TEXT="#000000">\n<H
    1>Unauthorized</H1>\n<H2>Error 401</H2>\n</BODY>\n</HTML>\n'

The first line in the tracebook is displaying the last two items that are part of the worksheet.append_row command. There are several items getting appended, so I don't know if it's just not including all of the items or it's just the end of the command causing the problem.

**Comment by bwatkin79** [comment]
@rhemon Where does your line "refresh_auth" come from? I'm new to this and trying to use your technique to fix this issue. Did you add the script by bryzgaloff as well or just the error exception? I have a good idea of where I'd need to put your script, but I don't know if there's a certain spot where bryzgaloff's script would go if necessary.

**Comment by bwatkin79** [comment]
@bryzgaloff I saw your possible solution to this HTTPError 401. I've never done anything like you wrote in your solution, so I have a couple questions. 

It looks like this goes in the beginning of my program in place of where I was authenticating before. Correct?

You say 'proxy class by worksheet'. Are you creating WKS to act in place of worksheet? If I use your script, does this mean that I'd make these changes:

worksheet.update_cell(1, 7, 'text') would become WKS.update_cell(1, 7, 'text')
worksheet.append_row() would become WKS.append_row()

I substituted my json file name and had this error occur. My console was pointing at the end of the file name:

    File "/home/bwatkin79/fangraphstest/fangraphsxpath.py", line 15
        self.credentials = ServiceAccountCredentials.from_json_keyfile_name(secret.Rotowire-MLB-0e0b82707568.json, scope)
                                                                                                       ^
    SyntaxError: invalid syntax

Please help. I'm trying to see if your solution will work for me too.

**Comment by rhemon** [comment]
@bwatkin79 refreh_auth is a function that i defined which just authorizes and opens the spreadsheet fine again, same thing bryzgaloff has done in his _refresh_auth method

**Comment by rhemon** [comment]
and it seems the solution works, but still sometimes i get the httperror but right after that if i retry the request, it works, so its kinda working for me

**Comment by bwatkin79** [comment]
@rhemon Thanks for replying. I'm totally new to this kind of method, but I have been doing some research. It looks to me like these solutions allow the program to continue populating with multiple API requests over a longer period of time. I've seen several people say that a batch update gets around this issue. I asked one person about this and the person believed a batch update wasn't supported by gspread. Have you explored that as a solution?

**Comment by bryzgaloff** [comment]
@bwatkin79 sorry, for late answer. Here is instruction:
1. You have to define my WKS class as a usual Python class to use it then in your code.
2. Next, create an instance of the WKS class like that:
```python
worksheet = WKS()
```
Please notice that the secret key for opening the spreadsheet in my code is hardcoded in `secret` package as `secret.GOOGLE_JSON_FILENAME`. So you have to create the same structure or, for instance, re-implement my `__init__` method and make the key to be a parameter of the function.
3. After all you can now use created `worksheet` object as a replacement for your previous `worksheet` instance of `gspread.Worksheet` class.

Your mistake was to use WKS class instead of its instance. That's why you have got an error, because the class itself doesn't have attribute `update_cell` but an instance of it does.

**Comment by euamotubaina** [comment]
Hi, thanks @bryzgaloff for the help

I'm trying modifying your code a bit but now I'm getting None returned

`import gspread
from oauth2client.service_account import ServiceAccountCredentials


class WKS:
    def __init__(self):
        scope = ['https://spreadsheets.google.com/feeds']
        self.credentials = ServiceAccountCredentials.from_json_keyfile_name('json.json', scope)
        self.spreadsheet_key = 'key'
        self.worksheet_name = 'Sheet'
        self._refresh_auth()

    def _refresh_auth(self):
        gc = gspread.authorize(self.credentials)
        self.wks = gc.open_by_key(self.spreadsheet_key).worksheet(self.worksheet_name)

    def _decorate(self, method):
        def safe_method(*args, **kwargs):
            try:
                method(*args, **kwargs)
            except gspread.exceptions.HTTPError as e:
                # getattr is needed to get a new instance of self.wks
                getattr(self.wks, method.__name__)(*args, **kwargs)
        return safe_method

    def __getattr__(self, attr):  # doesn't shadow _refresh_auth and _decorate
        return self._decorate(getattr(self.wks, attr))


worksheet = WKS()

data = worksheet.range(1, 1, 5, 2)

print(data)`

I'm really new to Python, I'm still trying to wrap my head around classes and instances.

thanks!

**Comment by kotborealis** [comment]
There's a little mistake in @bryzgaloff code; basically, there's ``return``s missing in ``safe_method``.

This should work:

```python
import gspread
from oauth2client.service_account import ServiceAccountCredentials

class WKS:
    def __init__(self):
        scope = 'https://www.googleapis.com/auth/spreadsheets'
        self.credentials = ServiceAccountCredentials.from_json_keyfile_name('./client_secret.json', scope)
        self._refresh_auth()

    def _refresh_auth(self):
        gc = gspread.authorize(self.credentials)
        self.wks = gc.open_by_key(config["gsheet_key"]).sheet1

    def _decorate(self, method):
        def safe_method(*args, **kwargs):
            try:
                return method(*args, **kwargs)
            except gspread.exceptions.HTTPError as e:
                # getattr is needed to get a new instance of self.wks
                return getattr(self.wks, method.__name__)(*args, **kwargs)
        return safe_method

    def __getattr__(self, attr):  # doesn't shadow _refresh_auth and _decorate
        return self._decorate(getattr(self.wks, attr))
```

**Comment by bryzgaloff** [comment]
@kotborealis thanks for your fix, I've updated my comment

**Comment by burnash** [comment]
This should be fixed with #637. On PyPI since version 3.4.0.

I'm closing this issue. Please upgrade gspread. If the problem persists, ping me and I'll reopen the issue.
